toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -110,13 +119,85 @@ func fetchPackages() (err error) {
 		return
 	}
 
+	// Optional delta build cache hydration
+	if *tryDownloadDeltaRPMs {
+		logger.Log.Info("Attempting to download delta RPMs for build nodes")
+		err = downloadDeltaNodes(dependencyGraph, cloner)
+		if err != nil {
+			err = fmt.Errorf("failed to download delta RPMs: %w", err)
+			return
+		}
+		// Update the package graph with the paths to the delta RPMs we downloaded
+		err = pkggraph.WriteDOTGraphFile(dependencyGraph, *outputGraph)
+		if err != nil {
+			err = fmt.Errorf("failed to write cache graph to file: %w", err)
+			return
+		}
+	}
+
 	// If we grabbed any RPMs, we need to convert them into a local repo
-	if hasUnresolvedNodes {
+	if *tryDownloadDeltaRPMs || hasUnresolvedNodes {
+		logger.Log.Info("Configuring downloaded RPMs as a local repository")
 		err = cloner.ConvertDownloadedPackagesIntoRepo()
 		if err != nil {
 			err = fmt.Errorf("failed to convert downloaded RPMs into a repo:\n%w", err)
 			return err
 		}
+
+		if strings.TrimSpace(*outputSummaryFile) != "" {
+			err = repoutils.SaveClonedRepoContents(cloner, *outputSummaryFile)
+			if err != nil {
+				logger.Log.Errorf("Failed to save cloned repo contents.")
+				return
+			}
+		}
+	}
+	return
+}
+
+// downloadDeltaNodes will look at the final cached graph we saved and see if any RPMS can be download instead of built.
+// If the previous part of the fetcher worked well we should be able to download only the delta RPMs we need
+// to build our packages or image (i.e. we should be able to create a subgraph just like we would for the build step)
+//   - dependencyGraph: The graph to use to find the packages we need to build. Should have any caching operations already
+//     performed on it. Will be updated with the paths to the delta RPMs we download.
+//   - cloner: The cloner to use to download the RPMs
+func downloadDeltaNodes(dependencyGraph *pkggraph.PkgGraph, cloner *rpmrepocloner.RpmRepoCloner) (err error) {
+	const (
+		deltaBuildGraph = true
+	)
+
+	timestamp.StartEvent("delta package download", nil)
+	defer timestamp.StopEvent(nil)
```

Comment:
>timestamp.StopEvent(nil) [](http://example.com/codeflow?start=7&length=24)

Curious: do you happen to know how `StopEvent` works? It seems we almost always pass `nil` as its argument, so how does it know which even to stop?

Reply:
There is a stack of timestamp events, we always stop the most recent one that is running. This way you can just use `StartEvent()` followed by `defer StopEvent()` and it will work even if more timestamps are created by subsequent calls.

If a single parent needs to start mutiple parallel child events then we can pass around the exact pointers and override that behavior.